File: scala/src/main/scala/io/devorbit/sparklabs/batch/sales/rdd/OrdersRevenue.scala
package io.devorbit.sparklabs.batch.sales.rdd

object OrdersRevenue extends App {
  val productPriceById = RddUtils.products.map(product => (product.product_id, product.price))

  RddUtils.orderItems
    .map(item => (item.product_id, item))
    .join(productPriceById)
    .map { case (_, (item, price)) => (item.order_id, (price * item.quantity, 1)) }
    .reduceByKey { case ((revenueA, countA), (revenueB, countB)) => (revenueA + revenueB, countA + countB) }
    .sortBy(_._2._1, ascending = false)
    .collect()
    .foreach(println)
}

File: python/sparklabs/batch/sales/rdd/orders_revenue.py
from sparklabs.batch.sales.rdd import utils


def run() -> None:
    product_price_by_id = utils.products.map(
        lambda product: (product["product_id"], product["price"])
    )
    rows = (
        utils.order_items.map(lambda item: (item["product_id"], item))
        .join(product_price_by_id)
        .map(lambda item: (item[1][0]["order_id"], (item[1][1] * item[1][0]["quantity"], 1)))
        .reduceByKey(lambda left, right: (left[0] + right[0], left[1] + right[1]))
        .sortBy(lambda item: item[1][0], ascending=False)
        .collect()
    )
    for row in rows:
        print(row)


if __name__ == "__main__":
    run()
